core/src/main/java/org/sonarsource/sonarlint/core/SonarLintClientImpl.java

```diff
@@ -19,126 +19,242 @@
+    } finally {
+      rwl.readLock().unlock();
     }
   }
 
-  private static RuntimeException handleException(RuntimeException t) {
-    if (LOG.isDebugEnabled()) {
-      // In DEBUG mode always return full exception
-      return convertToSonarLintException(t);
+  @Override
+  public void sync(ServerConfiguration serverConfig) {
+    checkNotNull(serverConfig);
+    checkConnectedMode();
+    rwl.writeLock().lock();
+    stop();
+    changeState(State.SYNCING);
+    ConnectedContainer connectedContainer = new ConnectedContainer(globalConfig, serverConfig);
+    try {
+      connectedContainer.startComponents();
+      connectedContainer.sync();
+    } catch (RuntimeException e) {
+      throw SonarLintWrappedException.build(e);
+    } finally {
+      try {
+        connectedContainer.stopComponents(false);
+      } catch (Exception e) {
+        // Ignore
+      }
+      start();
+      rwl.writeLock().unlock();
     }
+  }
 
-    for (Throwable y : Throwables.getCausalChain(t)) {
-      if (y instanceof MessageException) {
-        return convertToSonarLintException(y);
+  private void checkConnectedMode() {
+    if (globalConfig.getServerId() == null) {
+      throw new UnsupportedOperationException("Requires to be in connected mode");
+    }
+  }
+
+  @Override
+  public ValidationResult validateCredentials(ServerConfiguration serverConfig) {
+    checkNotNull(serverConfig);
+    checkConnectedMode();
+    rwl.readLock().lock();
+    ConnectedContainer connectedContainer = new ConnectedContainer(globalConfig, serverConfig);
+    try {
+      connectedContainer.startComponents();
+      return connectedContainer.validateCredentials();
+    } catch (RuntimeException e) {
+      throw SonarLintWrappedException.build(e);
+    } finally {
+      try {
+        connectedContainer.stopComponents(false);
+      } catch (Exception e) {
+        // Ignore
       }
+      rwl.readLock().unlock();
+    }
+  }
+
+  @Override
+  public List<RemoteModule> searchModule(String searchString) {
+    checkConnectedMode();
+    rwl.readLock().lock();
+    try {
+      return ((StorageGlobalContainer) globalContainer).searchModule(searchString);
+    } catch (RuntimeException e) {
+      throw SonarLintWrappedException.build(e);
+    } finally {
+      rwl.readLock().unlock();
     }
+  }
 
-    return convertToSonarLintException(t);
+  @Override
+  public void syncModule(ServerConfiguration serverConfig, String moduleKey) {
+    checkNotNull(serverConfig);
+    checkConnectedMode();
+    rwl.writeLock().lock();
+    stop();
+    changeState(State.SYNCING);
+    ConnectedContainer connectedContainer = new ConnectedContainer(globalConfig, serverConfig);
+    try {
+      connectedContainer.startComponents();
+      connectedContainer.syncModule(moduleKey);
+    } catch (RuntimeException e) {
+      throw SonarLintWrappedException.build(e);
+    } finally {
+      try {
+        connectedContainer.stopComponents(false);
+      } catch (Exception e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)